### DISCUSSION TITLE
[help] tar_stan_mcmc() doesn't accept target for stan_files argument?

### DISCUSSION THREAD

**Discussion by AdamChapnik1** [wlandau/stantargets#67] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I'm aware that this is an unusual use case but I have a {targets} pipeline that generates Stan code files that can change depending on some specification set dynamically in the pipeline. As a result, I have a target in the pipeline that is the path to the Stan file, but `tar_stan_mcmc()` immediately throws an error indicating it doesn't recognize that the object supplied for the `stan_files` argument is a target with the path to the Stan file as a character string. I was hoping to not have to define my own functions (or borrow from the [targets-stan](https://github.com/wlandau/targets-stan) repo) to run the Stan model, so I am wondering if there is a simple fix to this? Thank you!

Here's an example.

```R
# _targets.R
library(targets)
library(stantargets)

write_stan <- function(dir) {
  
  lines <- "data {
    int <lower = 1> n;
    vector[n] x;
    vector[n] y;
    real true_beta;
  }
  parameters {
    real beta;
  }
  model {
    y ~ normal(x * beta, 1);
    beta ~ normal(0, 1);
  }"
  cmdstanr::write_stan_file(lines, dir)
  
}

generate_data <- function(n = 10) {
  true_beta <- stats::rnorm(n = 1, mean = 0, sd = 1)
  x <- seq(from = -1, to = 1, length.out = n)
  y <- stats::rnorm(n, x * true_beta, 1)
  list(n = n, x = x, y = y, true_beta = true_beta)
}

# The _targets.R file ends with a list of target objects
# produced by stantargets::tar_stan_mcmc(), targets::tar_target(), or similar.
list(
  tar_target(path, write_stan(tempdir()), format = "file"),
  tar_stan_mcmc(
    example,
    stan_files = path,
    generate_data(),
    stdout = R.utils::nullfile(),
    stderr = R.utils::nullfile()
  )
)
```

The error you should get upon sourcing `_targets.R` is `Error in eval(ei, envir) : object 'path' not found`.

**Comment by wlandau** [comment]
`tar_stan_mcmc()` [statically branches](https://books.ropensci.org/targets/static.html) over Stan files, which means they need to be specified in advance. If the computation that generates the Stan files is quick, then you can call `write_stan()` to generate a global object and then use that as input for `tar_stan_mcmc()`.

```r
# _targets.R
library(targets)
library(stantargets)

write_stan <- function(dir) {
  
  lines <- "data {
    int <lower = 1> n;
    vector[n] x;
    vector[n] y;
    real true_beta;
  }
  parameters {
    real beta;
  }
  model {
    y ~ normal(x * beta, 1);
    beta ~ normal(0, 1);
  }"
  cmdstanr::write_stan_file(lines, dir)
  
}

generate_data <- function(n = 10) {
  true_beta <- stats::rnorm(n = 1, mean = 0, sd = 1)
  x <- seq(from = -1, to = 1, length.out = n)
  y <- stats::rnorm(n, x * true_beta, 1)
  list(n = n, x = x, y = y, true_beta = true_beta)
}

stan_file <- write_stan("persistent_path.stan")

list(
  tar_stan_mcmc(
    example,
    stan_files = stan_file,
    generate_data(),
    stdout = R.utils::nullfile(),
    stderr = R.utils::nullfile()
  )
)
```

**Reply by AdamChapnik1** [reply]
Thanks for the clarification! As a global object this works for me, but I tested and it seems like `tar_stan_mcmc()` doesn't work with `tar_map()`, unless I'm doing something wrong? 

```R
stan_files <- tibble(
  dirs = c("persistent_path_1", "persistent_path_2")
                     ) %>% 
  mutate(stan_file = map(dirs, write_stan) %>% list_c())

targets <- tar_map(
  values = stan_files,
  names = "dirs",
  # tar_target(example, dirs)
  tar_stan_mcmc(
    example,
    stan_files = dirs,
    generate_data(),
    stdout = R.utils::nullfile(),
    stderr = R.utils::nullfile()
  )
)
list(targets)
```

Anyway, I hadn't noticed but I now see that `tar_stan_mcmc()` also can statically branch over _multiple_ models, although each model will run on the same dataset. Unfortunately this still doesn't quite serve as a solution for my use case, and I think maybe the example code I gave initially didn't quite get at the use case I have in mind. 

As I mentioned, it is unusual but I am trying to build a pipeline that dynamically produces multiple different datasets from the same source data (basically just different sampling frameworks) and with that dynamically produces a different model for each dataset produced (because the set of data parameters will vary for each model for reasons it's not worth getting into). With those two target lists of 1-1 correspondence (of data and Stan models) I'd ideally only need to call `tar_stan_mcmc()` once, along with a `pattern` and `iteration` argument.

If `tar_stan_mcmc()` does in fact statically branch over Stan files in `tar_map()`, I can imagine a solution here involving static branching following the dynamic branching I've described, but otherwise I'm a bit stuck. Do you have any guidance on this or is what I'm asking just not feasible right now?

**Comment by wlandau** [comment]
> tar_stan_mcmc() doesn't work with tar_map(), unless I'm doing something wrong?

`tar_stan_mcmc()` already uses `tar_map()`, so adding one on top of the other might be hard. If that's all you need, `tarchetypes::tar_eval()` instead.

> Do you have any guidance on this or is what I'm asking just not feasible right now?

I don't think there would be room in `stantargets` for something so custom. You may have better luck using dynamic and static branching directly with the functions in `targets` and `tarchetypes`.
